samples/xtensa_asm2/src/main.c

```diff
@@ -0,0 +1,539 @@
+		printk("  WINDOWBASE %d -> %d, WINDOWSTART 0x%x -> 0x%x (%d cycles)\n",
+		       spill_wb0, spill_wb1, spill_ws0, spill_ws1,
+		       spill_end - spill_start);
+
+		if (spill_mode == NO_SPILL) {
+			spill_expect = result;
+			continue;
+		}
+
+		if (spill_ws1 != 1 << spill_wb1) {
+			printk("WINDOWSTART should show exactly one frame at WINDOWBASE\n");
+			ok = 0;
+		}
+
+		if (result != spill_expect) {
+			printk("Unexpected fn(1, 2, 3) result, got %d want %d\n",
+			       result, spill_expect);
+			ok = 0;
+		}
+	}
+
+	return ok;
+}
+
+int *test_highreg_handle;
+
+/* Simple save locations for some context needed by the test assembly */
+void *_test_highreg_sp_save;
+void *_test_highreg_a0_save;
+
+int test_highreg_stack[64];
+
+int *test_highreg_sp_top = &test_highreg_stack[ARRAY_SIZE(test_highreg_stack)];
+
+/* External function, defined in assembly */
+void fill_window(void (*fn)(void));
+
+/* Test rig for fill_window, maybe remove as a metatest */
+int testfw_wb, testfw_ws;
+void testfw(void);
+
+/* Assembly-defined leaf functions for fill_window which poke the
+ * specified number of high GPRs before calling xtensa_save_high_regs
+ * to spill them into the test_highreg_stack area for inspection.
+ */
+void test_highreg_0(void);
+void test_highreg_4(void);
+void test_highreg_8(void);
+void test_highreg_12(void);
+
+typedef void (*test_fn_t)(void);
+test_fn_t highreg_tests[] = {
+	test_highreg_0,
+	test_highreg_4,
+	test_highreg_8,
+	test_highreg_12,
+};
+
+int test_highreg_save(void)
+{
+	int ok = 1;
+
+	fill_window(testfw);
+	printk("testfw wb %d ws 0x%x\n", testfw_wb, testfw_ws);
+	ok = ok && (testfw_ws == ((1 << (XCHAL_NUM_AREGS / 4)) - 1));
+
+	for (int i = 0; i < ARRAY_SIZE(highreg_tests); i++) {
+		printk("\nHighreg test %d\n", i);
+
+		fill_window(highreg_tests[i]);
+
+		ok = ok && (*test_highreg_handle == (int)test_highreg_sp_top);
+
+		int spilled_words = test_highreg_sp_top - test_highreg_handle;
+
+		for (int quad = 0; ok && quad < (spilled_words - 1)/4; quad++) {
+			int *qbase = test_highreg_sp_top - (quad + 1) * 4;
+
+			for (int ri = 0; ri < 4; ri++) {
+				int reg = 4 + quad * 4 + ri;
+
+				ok = ok && (qbase[ri] == reg);
+				printk("  q %d reg %d qb[%d] %d\n",
+				       quad, reg, ri, qbase[ri]);
+			}
+		}
+	}
+
+	return ok;
+}
+
+void *switch_handle0, *switch_handle1;
+
+void xtensa_switch(void *handle, void **old_handle);
+
+void test_switch_bounce(void);
+__asm__("test_switch_bounce:"	"\n\t"
+	"call4 test_switch_top"	"\n\t");
+
+volatile int switch_count;
+
+/* Sits in a loop switching back to handle0 (which is the main thread) */
+void test_switch_top(void)
+{
+	int n = 1;
+
+	while (1) {
+		switch_count = n++;
+		xtensa_switch(switch_handle0, &switch_handle1);
+	}
+}
+
+int test_switch(void)
+{
+	static int stack2[512];
+
+	printk("%s\n", __func__);
+
+	memset(stack2, 0, sizeof(stack2));
+
+	int *sp = xtensa_init_stack(&stack2[ARRAY_SIZE(stack2)],
+				    (void *)test_switch_bounce,
+				    0, 0, 0);
+
+#if 0
```

Comment:
consider getting rid of this unused code block, or conditionally compiling with some debug Kconfig instead of #if 0.
If you really want it set up this way it's fine w/me, more of a suggestion.

Reply:
this code block should be removed.